src/service/resources-impl.js

```diff
@@ -1454,15 +1454,19 @@ export class Resources {
           this.exec_.getSize() == 0 &&
           this.queue_.getSize() == 0 &&
           now > this.exec_.getLastDequeueTime() + 5000) {
-      let idleScheduledCount = 0;
+      let idleScheduledPreloadCount = 0;
+      let idleScheduledLayoutCount = 0;
       for (let i = 0; i < this.resources_.length; i++) {
         const r = this.resources_[i];
         if (r.getState() == ResourceState.READY_FOR_LAYOUT &&
                 !r.hasOwner() && r.isDisplayed()) {
-          dev().fine(TAG_, 'idle layout:', r.debugid);
-          this.scheduleLayoutOrPreload_(r, /* layout */ false);
-          idleScheduledCount++;
-          if (idleScheduledCount >= 4) {
+          const idleLayout = r.idleRenderOutsideViewport();
+          if ((idleLayout && idleScheduledLayoutCount++ < 4) ||
```

Comment:
Ok, reviewing this, I think we need a Phase 6. Phase 5 and 6 will share an `idleScheduledCount`.

Phase 5 will be trying to find elements that haven't scheduled yet and are idle render eligible. These will schedule _preload_. Once we hit 4, break.

If Phase 5 completes and we're under the 4 limit, we can iterate Phase 6 and just schedule things that aren't until we hit the 4 limit.

Reply:
I'm concerned that an abundance of preloads will cause layout on idle to never execute (for instance amp-ads themselves include preload).  Preload are inherently network based while layout can be CPU intensive.  My hope is that they're exclusive enough to be independent.

Reply:
Done